sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java

```diff
@@ -0,0 +1,203 @@
+import org.eclipse.paho.client.mqttv3.MqttException;
+import org.eclipse.paho.client.mqttv3.MqttMessage;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Tests of {@link MqttIO}.
+ */
+@RunWith(JUnit4.class)
+public class MqttIOTest implements Serializable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MqttIOTest.class);
+
+  private static transient BrokerService broker;
+
+  private static int port;
+
+  @BeforeClass
+  public static void startBroker() throws Exception {
+    LOGGER.info("Finding free network port");
+    ServerSocket socket = new ServerSocket(0);
+    port = socket.getLocalPort();
+    socket.close();
+
+    LOGGER.info("Starting ActiveMQ broker on {}", port);
+    broker = new BrokerService();
+    broker.setUseJmx(false);
+    broker.setPersistenceAdapter(new MemoryPersistenceAdapter());
+    broker.addConnector(new URI("mqtt://localhost:" + port));
+    broker.start();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testRead() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+
+    PCollection<byte[]> output = pipeline.apply(
+        MqttIO.read()
+            .withConnectionConfiguration(
+                MqttIO.ConnectionConfiguration.create(
+                    "tcp://localhost:" + port,
+                    "BEAM_PIPELINE",
+                    "READ_TOPIC"))
+          .withMaxNumRecords(10));
+
+    PAssert.thatSingleton(output.apply("Count", Count.<byte[]>globally()))
+        .isEqualTo(10L);
+    PAssert.that(output).satisfies(new SerializableFunction<Iterable<byte[]>, Void>() {
+      @Override
+      public Void apply(Iterable<byte[]> input) {
+        for (byte[] element : input) {
+          String inputString = new String(element);
+          Assert.assertTrue(inputString.startsWith("This is test "));
+          int count = Integer.parseInt(inputString.substring("This is test ".length()));
+          Assert.assertTrue(count < 10);
+          Assert.assertTrue(count >= 0);
+        }
+        return null;
+      }
+    });
+
+    // produce messages on the broker in another thread
+    // This thread prevents to block the pipeline waiting for new messages
+    Thread thread = new Thread() {
+      public void run() {
+        try {
+          // gives time to the pipeline to start
+          Thread.sleep(2000);
+        } catch (Exception e) {
+          // nothing to do
+        }
+        try {
+          MqttClient client = new MqttClient("tcp://localhost:" + port, "publisher");
+          client.connect();
+          for (int i = 0; i < 10; i++) {
+            MqttMessage message = new MqttMessage();
+            message.setQos(1);
+            message.setRetained(true);
```

Comment:
The way I understand it, this will overwrite "the" retained message on the topic 10 times (i.e. the retained message will be message number 9), rather than queue up 10 retained messages. I'm very confused as to how this will interact with the reader thread.